File: sorting/c/selectionsort.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Implementation of the Selection Sort
 */

void selection_sort(int [], int);
void swap(int *, int *);
int find_minimum(int [], int, int);
void print_array(int [], int);

int
main(int argc, char **argv) {
	int array[20] = {
		 9, 71,  2, 84, 12, 23, 99, 26,  1, 76,
		67,  4, 53, 42, 10, 38, 66,  7, 80, 2
	};

	print_array(array, 20);
	selection_sort(array, 20);
	print_array(array, 20);

	return EXIT_SUCCESS;
};

/*
 * Print the array
 */
void
print_array(int array[], int length) {
	int i;
	for (i = 0; i < length; i++)
		printf("%3d ", array[i]);
	printf("\n");
}

/*
 * Sort the (integer) array with selection sort
 */
void
selection_sort(int array[], int length) {
	int i, min;

	for (i = 0; i < length; i++) {
		min = find_minimum(array, i, length);
		swap(&array[min], &array[i]);
	}
}

/*
 * Returns the index of the mininum element of the array within
 * the specified positions
 */
int
find_minimum(int array[], int start_pos, int end_pos) {
	int min = start_pos, pos = start_pos;

	while (pos < end_pos) {
		if (array[pos] < array[min])
			min = pos;
		pos++;
	}

	return min;
}

/*
 * Swap two items
 */
void
swap(int *item1, int *item2) {
	int tmp = *item1;
	*item1 = *item2;
	*item2 = tmp;
}

File: sorting/java/SelectionSort.java
package sorting.java;

import java.util.Arrays;

/**
 * Implementation of selection sort
 */
public class SelectionSort {
  public static <T extends Comparable<T>> T[] selectionsort(T[] items) {
    int size = items.length;
    int i, j, min;
    T tmp;

    for (i = 0; i < size; i++) {
      min = i;

      for (j = i + 1; j < size; j++) {
        if (items[j].compareTo(items[min]) < 0) {
          min = j;
        }
      }

      tmp = items[i];
      items[i] = items[min];
      items[min] = tmp;
    }

    return items;
  }

  public static void main(String[] args) {
    System.out.println(Arrays.toString(selectionsort(new Integer[]{1, 2, 3, 4, 5, 6, 7})));
    System.out.println(Arrays.toString(selectionsort(new Integer[]{5, 1, 6, 3, 4, 2, 7})));
    System.out.println(Arrays.toString(selectionsort(new Integer[]{7, 6, 5, 4, 3, 2, 1})));

    System.out.println(Arrays.toString(selectionsort(new String[]{"a", "b", "c", "d", "e"})));
    System.out.println(Arrays.toString(selectionsort(new String[]{"b", "c", "a", "e", "d"})));
    System.out.println(Arrays.toString(selectionsort(new String[]{"e", "d", "c", "b", "a"})));
  }
}
